### PR TITLE
feat(fw_mode_manager): Do first-order-hold also for go-to

### PR DESCRIPTION
**Pull request by sfuhrer**
## Solved Problem

When a fixed-wing vehicle in **Hold** receives a new **Go-to / reposition** setpoint at a different altitude, the navigator emits a `LOITER` setpoint carrying the new lat/lon **and** altitude. The fixed-wing loiter controller commanded the target altitude *immediately*, so the aircraft climbed/descended at its maximum rate up front and then flew level the rest of the way to the loiter point — a "steep ramp, then flat" profile.

An altitude **first-order hold (FOH)** already existed for plain mission waypoints (`SETPOINT_TYPE_POSITION`) to spread the altitude change smoothly over the leg, but it was **not** applied to `LOITER` setpoints, so Go-to / reposition (and loiter legs generally) didn't benefit from it.

## Solution

Apply the same altitude FOH in the loiter path. The altitude setpoint is now interpolated linearly with horizontal distance so the target altitude is reached at the acceptance radius / loiter circle, giving a single **diagonal** glide path instead of a steep ramp followed by level flight.

While at it, the FOH logic — previously inlined in `control_auto_position` — was extracted into a standalone, unit-tested function (`calculateFirstOrderHoldAltitude`) shared by both the position and loiter paths.

### Altitude profile: before vs. now

foh_altitude_profile

*Example: a Go-to in Hold to a point 1000 m away that is 100 m higher, loiter radius 200 m.*
- **Before (red):** command target altitude immediately → climb at max rate, then fly level.
- **Now (green):** ramp altitude over the leg → reach target altitude at the loiter circle (diagonal).

The behavior is symmetric for descents.

## Changes

- Extract the FOH altitude computation out of `FixedWingModeManager` into a standalone, unit-tested helper (`FirstOrderHoldAltitude.{cpp,hpp}`) with an explicit `FirstOrderHoldAltitudeState`.
- Apply the FOH ramp on the **loiter/go-to** path as well as position waypoints, so a go-to climbs/descends gradually along the transit and reaches the target altitude at the loiter circle.
- Rework the ramp so it is anchored on the **current commanded altitude** at the moment the target altitude changes (rather than interpolating between the previous and current waypoint altitudes), reaching the target at the acceptance radius / loiter circle. Progress is measured by the closest horizontal approach to the target.
- Reset the ramp state whenever the FOH did not run on the previous cycle (any non-FOH mode — manual, takeoff, landing, velocity, external nav, …), so re-engaging auto starts a fresh ramp from the current altitude instead of resuming stale state.

## Test coverage

Unit tests added (`make tests TESTFILTER=FixedWingModeManager`). The test links only the extracted translation unit, so it doesn't pull in the module's app/daemon registration.

## Discussion / things to confirm
For me the biggest controversy atm is whether the ramp should start from a measurement or previous setpoint. I opted for measurement for now (see 26e1206c2cbce6688c0ef829477f8d183c822935) . That way we don't need special logic to catch large altitude errors in the moment a new altitude setpoint is received. Drawback: the altitude reference is not continuous, it jumps to the current value.
image

## Test setup / how to verify in SITL

1. Fixed-wing SITL, take off and switch to **Hold**.
2. Command a **Go-to** to a location at a clearly different altitude.
3. Observe the altitude vs. ground-track profile: the vehicle should descend/climb diagonally toward the target rather than changing altitude up front and then flying level.

### CODE REVIEW

**Comment by sfuhrer** [comment]
@mahima-yoga @ryanjAA @Jaeyoung-Lim curious for feedback here. Have a look at the plot above: do you prefer the current (red) or proposed new (green) approach for altitude handling on FW?
No need to do code review yet.

**Comment by ryanjAA** [comment]
green - saves battery. less battery stress. less management if at high alt. safer overall. 

that said there are reasons to climb right now but thats handled in other levers (i.e. change alt)

**Review by Jaeyoung-Lim — COMMENTED** [review]
@sfuhrer I would say green is closer to the expected behavior, although if you are flying in terrain red is safer. I have had this conversation a few times by now on people being surprised that it climbs first. So I think it makes sense for us to change it to first order hold.

**Comment by mahima-yoga** [comment]
I also agree that green is likely what any new user would expect to happen. Same as Jay - I have also had some conversations with people that are confused with the red behaviour (and don't like it).

**Comment by Jaeyoung-Lim** [comment]
Putting a little bit more thought to it, still think green behavior is mor e expexted. However, I think it should still climb first if the altitude is not reachable within the max climbrate. Otherwise a misplanned mission would be more catastrophic

**Comment by sfuhrer** [comment]
@ryanjAA @Jaeyoung-Lim @mahima-yoga thanks for your feedback. Let's then do FOH also in Hold mode. I made a couple of adjustments to the code and updated the description, please review now!

File: src/modules/fw_mode_manager/FixedWingModeManager.cpp [hunk 1]
```diff
@@ -619,7 +620,7 @@ FixedWingModeManager::control_auto(const float control_interval, const Vector2d
 		} else
 #endif // CONFIG_FIGURE_OF_EIGHT
 		{
-			control_auto_loiter(control_interval, curr_pos, ground_speed, current_sp, pos_sp_next);
+			control_auto_loiter(control_interval, curr_pos, ground_speed, pos_sp_prev, current_sp, pos_sp_next);
```
Comment (mahima-yoga) [inline src/modules/fw_mode_manager/FixedWingModeManager.cpp]: I don't see `pos_sp_prev `being used in the function

File: src/modules/fw_mode_manager/FirstOrderHoldAltitude.cpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+#include <lib/mathlib/mathlib.h>
+
+float calculateFirstOrderHoldAltitude(const position_setpoint_s &pos_sp_curr, const double current_lat,
+				      const double current_lon, const float current_altitude, const float acc_rad,
+				      FirstOrderHoldAltitudeState &state)
+{
+	const float target_altitude = pos_sp_curr.alt;
+
+	// The current altitude is considered reached within the acceptance radius (or loiter circle) of the target.
+	const float completion_radius = math::max(acc_rad, fabsf(pos_sp_curr.loiter_radius));
```
Comment (mahima-yoga) [inline src/modules/fw_mode_manager/FirstOrderHoldAltitude.cpp]: Do you think this line is required? Since the acc_rad is already passed, and in the loiter case is already constrained. In the position case is it necessary?

File: src/modules/fw_mode_manager/FixedWingModeManager.cpp [hunk 3]
```diff
@@ -938,6 +910,13 @@ FixedWingModeManager::control_auto_loiter(const float control_interval, const Ve
 		loiter_direction_ccw = _param_nav_loiter_rad.get() < -FLT_EPSILON;
 	}
 
+	// Smoothly ramp the altitude setpoint from the previous setpoint such that the loiter altitude is reached
```
Comment (mahima-yoga) [inline src/modules/fw_mode_manager/FixedWingModeManager.cpp]:
```suggestion
	// Smoothly ramp the altitude setpoint from the current altitude such that the loiter altitude is reached
```

File: src/modules/fw_mode_manager/FirstOrderHoldAltitudeTest.cpp [hunk 4]
```diff
@@ -0,0 +1,179 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2026 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
```
Comment (mahima-yoga) [inline src/modules/fw_mode_manager/FirstOrderHoldAltitudeTest.cpp:5]: Not super opinionated on this but I would rename this file to `FirstOrderHoldAltitudeTest.cpp`

File: src/modules/fw_mode_manager/FirstOrderHoldAltitude.cpp (hunk 2 above)
Comment (mahima-yoga) [reply src/modules/fw_mode_manager/FirstOrderHoldAltitude.cpp]: If it's not required we could then pass just the `setpoint.lat/lon/alt` directly and drop the dependency on the `position_setpoint`.

**Review by mahima-yoga — COMMENTED** [review]
I think the implementation makes sense and I agree with ramping from the current altitude. Don't see any big issues in the implementation, just left a few small comments. 

But as we discussed in person: I think it would make sense to separate any "abort" functionality to prioritize altitude.

File: src/modules/fw_mode_manager/FirstOrderHoldAltitude.cpp (hunk 2 above)
Comment (mahima-yoga) [reply src/modules/fw_mode_manager/FirstOrderHoldAltitude.cpp]: Actually here should we not use `NAV_FW_ALT_RAD`

Comment (sfuhrer) [reply src/modules/fw_mode_manager/FirstOrderHoldAltitude.cpp]: Not sure I follow here:
- `pos_sp_curr.loiter_radius` is required such that if you set a big loiter, the vehicle always reaches the altitude the moment it starts loitering
- which dependency of `position_setpoint`? `pos_sp_curr` is also used in many other places inside this function.
- where is  `NAV_FW_ALT_RAD` used atm?

File: src/modules/fw_mode_manager/FirstOrderHoldAltitudeTest.cpp (hunk 4 above)
Comment (sfuhrer) [reply src/modules/fw_mode_manager/FirstOrderHoldAltitudeTest.cpp:5]: Thought also about it, would you then put other unit test for the FWModeManager inside other, dedicated files? (I know that currently we have none).

File: src/modules/fw_mode_manager/FirstOrderHoldAltitude.cpp (hunk 2 above)
Comment (mahima-yoga) [reply src/modules/fw_mode_manager/FirstOrderHoldAltitude.cpp]: - I meant that for the loiter-case, the `acc_rad` is [already set based on the loiter radius](https://github.com/PX4/PX4-Autopilot/pull/27747/changes#diff-0e624ccf7b4ffa7466c2bef13a4a129ea57adf4581ce1593102dedd1d8dd4826R915) before calling `calculateFirstOrderHoldAltitude()`.  

- If the function doesn't require loiter_radius, then you could pass the fields of the setpoint (lat, lon, alt) rather than the setpoint itself and it would be a standalone geometric function without the dependency on `position_setpoint` in the header file. (But that's a personal preference) 

- `NAV_FW_ALT_RAD` : This doesn't have any relevance here sorry, I got confused

File: src/modules/fw_mode_manager/FirstOrderHoldAltitudeTest.cpp (hunk 4 above)
Comment (mahima-yoga) [reply src/modules/fw_mode_manager/FirstOrderHoldAltitudeTest.cpp:5]: My thinking was that since `FirstOrderHoldAltitude` is self-contained, it's clean to have a dedicated test file. If we would later add higher-level tests for the manager itself, those could then go in a `FixedWingModeManagerTest.cpp`.

Comment (sfuhrer) [reply src/modules/fw_mode_manager/FirstOrderHoldAltitudeTest.cpp:5]: Okay you convinced me, pushed!

File: src/modules/fw_mode_manager/FirstOrderHoldAltitude.cpp (hunk 2 above)
Comment (sfuhrer) [reply src/modules/fw_mode_manager/FirstOrderHoldAltitude.cpp]: Got you, that makes a lot of sense. Pushed the change.

File: src/modules/fw_mode_manager/FirstOrderHoldAltitude.hpp [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#pragma once
+
+#include <float.h>
+#include <math.h>
+
+#include <uORB/topics/position_setpoint.h>
```
Comment (mahima-yoga) [inline src/modules/fw_mode_manager/FirstOrderHoldAltitude.hpp]:
```suggestion
```

**Review by mahima-yoga — DISMISSED** [review]
Looks good!

**Review by mahima-yoga — APPROVED** [review]
